packages/puppeteer-core/src/node/FirefoxLauncher.ts

```diff
@@ -110,86 +126,75 @@ export class FirefoxLauncher extends ProductLauncher {
     }
 
-    const runner = new BrowserRunner(
-      this.product,
-      firefoxExecutable,
-      firefoxArguments,
+    return {
+      isTempUserDataDir,
       userDataDir,
-      isTempUserDataDir
-    );
-    runner.start({
-      handleSIGHUP,
-      handleSIGTERM,
-      handleSIGINT,
-      dumpio,
-      env,
-      pipe,
-    });
+      args: firefoxArguments,
+      executablePath: firefoxExecutable,
+    };
+  }
 
-    if (protocol === 'webDriverBiDi') {
-      let browser: Browser;
+  /**
+   * @internal
+   */
+  override async cleanUserDataDir(
+    userDataDir: string,
+    opts: {isTemp: boolean}
+  ): Promise<void> {
+    if (opts.isTemp) {
       try {
-        const connection = await runner.setupWebDriverBiDiConnection({
-          timeout,
-          slowMo,
-          preferredRevision: this.puppeteer.browserRevision,
-          protocolTimeout,
-        });
-        const BiDi = await import(
-          /* webpackIgnore: true */ '../common/bidi/bidi.js'
-        );
-        browser = await BiDi.Browser.create({
-          connection,
-          closeCallback: runner.close.bind(runner),
-          process: runner.proc,
-        });
+        await rm(userDataDir);
       } catch (error) {
-        runner.kill();
+        debugError(error);
         throw error;
       }
-
-      return browser;
-    }
-
-    let browser;
-    try {
-      const connection = await runner.setupConnection({
-        usePipe: pipe,
-        timeout,
-        slowMo,
-        preferredRevision: this.puppeteer.browserRevision,
-        protocolTimeout,
-      });
-      browser = await CDPBrowser._create(
-        this.product,
-        connection,
-        [],
-        ignoreHTTPSErrors,
-        defaultViewport,
-        runner.proc,
-        runner.close.bind(runner),
-        options.targetFilter
-      );
-    } catch (error) {
-      runner.kill();
-      throw error;
+    } else {
+      try {
+        // When an existing user profile has been used remove the user
+        // preferences file and restore possibly backuped preferences.
+        await unlink(path.join(userDataDir, 'user.js'));
+
+        const prefsBackupPath = path.join(userDataDir, 'prefs.js.puppeteer');
+        if (fs.existsSync(prefsBackupPath)) {
+          const prefsPath = path.join(userDataDir, 'prefs.js');
+          await unlink(prefsPath);
+          await rename(prefsBackupPath, prefsPath);
+        }
+      } catch (error) {
+        debugError(error);
+      }
     }
+  }
 
-    if (waitForInitialPage) {
+  /**
+   * @internal
+   */
+  override async closeBrowser(
+    browserProcess: ReturnType<typeof launch>,
+    connection?: Connection
+  ): Promise<void> {
+    if (connection) {
+      // Attempt to close the browser gracefully
       try {
-        await browser.waitForTarget(
-          t => {
-            return t.type() === 'page';
-          },
-          {timeout}
-        );
+        await connection.send('Browser.close');
```

Comment:
This only works for CDP (My presumption is that CDP is case specific), the command will be different `browser.close` so they will be incomparable. Let's put a `TODO: BiDi https://github.com/w3c/webdriver-bidi/pull/286` comment
Also both methods are the same for ChromeLauncher and FirefoxLauncher, lets move it to ProductLauncher as well.